movie_search/middleware.py

```diff
@@ -0,0 +1,16 @@
+from .tmdb_api import TMDBApi
+from .media import MediaService
+from django.conf import settings
+
+class TMDBApiMiddleware:
+    def __init__(self, get_response):
+        self.get_response = get_response
+    
+    def __call__(self, request):
+        print("Calling before view......")
+        # Attach a configured TMDBApi instance to the request
+        request.tmdb_api = TMDBApi()
```

Comment:
this should be moved into the constructor actually, otherwise it makes an instance for each view call

> The code snippet you provided shows that the middleware is indeed designed to attach an instance of TMDBApi to every incoming request. However, the creation of the TMDBApi instance as described in your code happens within the __call__ method of the middleware:

Reply:
although moving it to the constructor will guarantee a single object, if you want to enforce it with a "singleton" ...

> If you want to enforce the singleton pattern directly within the TMDBApi class itself, rather than through middleware, you can modify the class to ensure a single instance is created. Here’s an example of how you might implement this pattern:

```
from django.conf import settings
import requests
from requests.exceptions import JSONDecodeError

class TMDBApi:
    _instance = None

    def __new__(cls, language=None):
        if cls._instance is None:
            cls._instance = super(TMDBApi, cls).__new__(cls)
            # Initialize with default values
            cls._instance.base_url = settings.TMDB_API_URL
            cls._instance.api_key = settings.TMDB_API_KEY
            cls._instance.language = language or settings.TMDB_API_LANG
        return cls._instance

    def get_data_from_endpoint(self, endpoint, **kwargs):
        url = f"{self.base_url}{endpoint}"
        params = {"api_key": self.api_key, "language": self.language}
        params.update(kwargs)
        response = requests.get(url, params=params)
        try:
            return response.json()
        except JSONDecodeError:
            # Handle JSON decoding error
            error_message = (f"Failed to parse JSON response for URL: {url} - "
                             f"Status Code: {response.status_code} - Response: {response.text[:500]}")
            raise JSONDecodeError(error_message)

    # Other methods...
```

But again might not be necessary, easiest is to just put an additional print in the constructor and see if it fires only once during the django app lifetime